pkg/p2p/libp2p/internal/reacher/reacher.go

```diff
@@ -0,0 +1,264 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package reacher runs a background worker that will ping peers
+// from an internal queue and report back the reachability to the notifier.
+package reacher
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/swarm"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	pingTimeout        = time.Second * 5
+	pingMaxAttempts    = 3
+	workers            = 8
+	retryAfterDuration = time.Second * 15
+)
+
+type peerState int
+
+const (
+	waiting peerState = iota
+	inProgress
+)
+
+type peer struct {
+	overlay    swarm.Address
+	addr       ma.Multiaddr
+	retryAfter time.Time
+	attempts   int
+	state      peerState
+}
+
+type reacher struct {
+	mu    sync.Mutex
+	peers map[string]*peer
+
+	work chan struct{}
+	quit chan struct{}
+
+	pinger   p2p.Pinger
+	notifier p2p.ReachableNotifier
+
+	wg      sync.WaitGroup
+	metrics metrics
+
+	options *Options
+}
+
+type Options struct {
+	PingTimeout        time.Duration
+	PingMaxAttempts    int
+	Workers            int
+	RetryAfterDuration time.Duration
+}
+
+func New(streamer p2p.Pinger, notifier p2p.ReachableNotifier, o *Options) *reacher {
+
+	r := &reacher{
+		work:     make(chan struct{}, 1),
+		quit:     make(chan struct{}),
+		pinger:   streamer,
+		peers:    make(map[string]*peer),
+		notifier: notifier,
+		metrics:  newMetrics(),
+	}
+
+	if o == nil {
+		o = &Options{
+			PingTimeout:        pingTimeout,
+			PingMaxAttempts:    pingMaxAttempts,
+			Workers:            workers,
+			RetryAfterDuration: retryAfterDuration,
+		}
+	}
+	r.options = o
+
+	r.wg.Add(1)
+	go r.manage()
+
+	return r
+}
+
+func (r *reacher) manage() {
+
+	defer r.wg.Done()
+
+	c := make(chan *peer)
+	defer close(c)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	r.wg.Add(r.options.Workers)
+	for i := 0; i < r.options.Workers; i++ {
+		go r.ping(c, ctx)
+	}
+
+	for {
+
+		p, tryAfter := r.tryAcquirePeer()
+
+		// if no peer is returned,
+		// wait until either more work or the closest retry-after time.
+
+		// wait for work and tryAfter
+		if tryAfter > 0 {
+			select {
+			case <-r.quit:
+				return
+			case <-r.work:
+				continue
+			case <-time.After(tryAfter):
+				continue
+			}
+		}
+
+		// wait for work
+		if p == nil {
+			select {
+			case <-r.quit:
+				return
+			case <-r.work:
+				continue
+			}
+		}
+
+		// send p to channel
+		select {
+		case <-r.quit:
+			return
+		case c <- p:
```

Comment:
just wondering whether it is possible that the same peer can be scheduled to multiple goroutines simultaneously, resulting in data races between them on the status of the peer